kura/org.eclipse.kura.driver.ble.xdk/src/main/java/org/eclipse/kura/internal/driver/ble/xdk/Xdk.java

```diff
@@ -0,0 +1,519 @@
+    private static final String SENSOR = "sensor";
+    private static final String RATE = "rate";
+    private static final String SENSOR_FUSION = "sensor fusion";
+    private static final String HIGH_PRIORITY_ARRAY = "high priority array";
+    private static final String LOW_PRIORITY_ARRAY = "low priority array";
+
+    private static final int SERVICE_TIMEOUT = 10000;
+
+    private byte[] value = { 0x01 };
+    private byte[] rate = { 0x64, 0x00, 0x00, 0x00 };
+    private byte[] sensorFusion = { 0x00 };
+
+    private BluetoothLeDevice device;
+    private Map<String, XdkGattResources> gattResources;
+
+    public Xdk(BluetoothLeDevice bluetoothLeDevice) {
+        this.device = bluetoothLeDevice;
+        this.gattResources = new HashMap<>();
+    }
+
+    public BluetoothLeDevice getBluetoothLeDevice() {
+        return this.device;
+    }
+
+    public void setBluetoothLeDevice(BluetoothLeDevice device) {
+        this.device = device;
+    }
+
+    public boolean isConnected() {
+        return this.device.isConnected();
+    }
+
+    public void connect() throws ConnectionException {
+        try {
+            this.device.connect();
+            // Wait a bit to ensure that the device is really connected and services discovered
+            Long startTime = System.currentTimeMillis();
+            while (System.currentTimeMillis() - startTime < SERVICE_TIMEOUT) {
+                if (this.device.isServicesResolved()) {
+                    break;
+                }
+                XdkDriver.waitFor(1000);
+            }
+            if (!isConnected() || !this.device.isServicesResolved()) {
+                throw new ConnectionException("Connection failed");
+            }
+        } catch (KuraBluetoothConnectionException e) {
+            throw new ConnectionException(e);
+        }
+
+    }
+
+    public void init() throws ConnectionException {
+        if (isConnected() && this.gattResources.size() != 8) {
+            getGattResources();
+        }
+    }
+
+    public void disconnect() throws ConnectionException {
+        if (isHighNotifying()) {
+            disableHighNotifications();
+        }
+        if (isLowNotifying()) {
+            disableLowNotifications();
+        }
+        try {
+            this.device.disconnect();
+            if (isConnected()) {
+                throw new ConnectionException("Disconnection failed");
+            }
+        } catch (KuraBluetoothConnectionException e) {
+            throw new ConnectionException(e);
+        }
+        // Wait a while after disconnection
+        XdkDriver.waitFor(1000);
+    }
+
+    public String getFirmareRevision() {
+        String firmware = "";
+        try {
+            BluetoothLeGattCharacteristic devinfo = this.gattResources.get(DEVINFO).getGattService()
+                    .findCharacteristic(XdkGatt.UUID_XDK_CONTROL_SERVICE_GET_FIRMWARE_VERSION);
+            firmware = new String(devinfo.readValue(), "UTF-8");
```

Comment:
You can change the specific UTF-8 string with the corresponding StandardCharsets UTF8 enum. This will also remove the catch for UnsupportedEncoding